### PR TITLE
kick patch

### PR DESCRIPTION
**Pull request by JesseAbram**
related to #87

### CODE REVIEW

File: pallets/collator-selection/src/benchmarking.rs [hunk 1]
```diff
@@ -134,21 +134,21 @@ benchmarks! {
...
-
 		T::Currency::make_free_balance_be(
 			&<CollatorSelection<T>>::account_id(),
-			T::Currency::minimum_balance() * 2u32.into(),
+			T::Currency::minimum_balance() * 4u32.into(),
 		);
 		let author = account("author", 0, SEED);
+		let new_block: T::BlockNumber = 10u32.into();
+
+		frame_system::Pallet::<T>::set_block_number(new_block.clone());
```
Comment (kianenigma) [inline pallets/collator-selection/src/benchmarking.rs:145]:
```suggestion
		frame_system::Pallet::<T>::set_block_number(new_block);
```

isn't block number `Copy`?

File: pallets/collator-selection/src/lib.rs [hunk 2]
```diff
@@ -353,11 +359,12 @@ pub mod pallet {
...
 		}
 		/// Kicks out and candidates that did not produce a block in the kick threshold.
-		pub fn kick_stale_candidates(candidates: Vec<CandidateInfo<T::AccountId, BalanceOf<T>, T::BlockNumber>>) -> Vec<T::AccountId> {
+		pub fn kick_stale_candidates(candidates: Vec<CandidateInfo<T::AccountId, BalanceOf<T>>>) -> Vec<T::AccountId> {
 			let now = frame_system::Pallet::<T>::block_number();
 			let kick_threshold = T::KickThreshold::get();
 			let new_candidates = candidates.into_iter().filter_map(|c| {
-				let since_last = now - c.last_block;
+				let last_block = <LastAuthoredBlock<T>>::get(c.who.clone());
+				let since_last = if now > last_block { now - last_block } else { 0u32.into() };
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:367]:
```suggestion
				let since_last = now.saturating_sub(last_block);
```

**Review by kianenigma — APPROVED** [review]
Some nits, but looks overall good to me.

**Comment by kianenigma** [comment]
As I read in the comment, this will not fix #87, right? what's the plan around that?

**Comment by JesseAbram** [comment]
> As I read in the comment, this will not fix #87, right? what's the plan around that?

I think it does, here it adds https://github.com/paritytech/statemint/pull/88/files#diff-a0359a0add814d66aac74493f7e17aefdcee21d3de1ae592e97059082b6feedfR314 kick threshold to the initial block. So if let's say a session is 100 blocks alice enters at 100 then here start block would be 200 and she would be able to produce a block in session 200-300 and not be kicked, I see it as a temp solution and we should eventually build out the Im online type suggestion you had before.

**Review by kianenigma — CHANGES_REQUESTED** [review]
Sorry, I thought a bit more, and want to discuss: Logic-wise, the change of this PR is that it is giving every onboarded collator an initial slack of `T::KickThreshold`. Once set to `Sessionlength`, this will have the desired effect, yet it is somewhat convoluted and the detail is hidden in the _implementation_, rather than being in the _interface_.

To achieve the same thing, I recommend just adding a comment that `type KickThreshold` should take into account session delay, whatever that may be. In our case, we want `Type KickThreshold = 2 * SessionLength`.

I am still okay with the other changes in the PR, such as making a new map, but that is not relevant as far as I can see.

**Review by kianenigma — APPROVED** [review]

**Comment by kianenigma** [comment]
> Sorry, I thought a bit more, and want to discuss: Logic-wise, the change of this PR is that it is giving every onboarded collator an initial slack of `T::KickThreshold`. Once set to `Sessionlength`, this will have the desired effect, yet it is somewhat convoluted and the detail is hidden in the _implementation_, rather than being in the _interface_.
> 
> To achieve the same thing, I recommend just adding a comment that `type KickThreshold` should take into account session delay, whatever that may be. In our case, we want `Type KickThreshold = 2 * SessionLength`.
> 
> I am still okay with the other changes in the PR, such as making a new map, but that is not relevant as far as I can see.

This is wrong and was consolidated offline.

**Pull request merged**